### PR TITLE
[WIP] Rewrite to use native Proxy

### PR DESCRIPTION
**Pull request by BryanCrotaz**
Work in progress. Using existing test suite plus some more edge cases and preserving public API.

Enables developer to use normal JS semantics:

```js
var output = changeset.gruffalos.lastObject.say('grrr');
changeset.people[1].name.first = 'Bryan';
```

Really useful for having nested components:
```hbs
{{#let (changeset this.model) as |changeset|}}
  {{#each changeset.people as | person |}}
    <PersonForm @model={{person}}/>
  {{/each}}
{{/let}}
```
PersonForm doesn't have to know that this is a changeset. The save and rollback can be at a higher level.

### CODE REVIEW

File: addon/classes/ChangesetArrayProxyHandler.ts [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+import IChangesetProxyHandler from './IChangesetProxyHandler';
+import ProxyOptions from './ProxyOptions';
+import { ProxyArrayValueKey } from 'ember-changeset/utils/symbols';
+
+type Input = any[] | ArrayProxy<any>;
+
+export default class ChangesetArrayProxyHandler<T extends Input> implements IChangesetProxyHandler<T> {
+  constructor(source: T, _options?: ProxyOptions) {
+    if (source?.constructor?.name === 'ArrayProxy') {
+      this.__sourceProxy = source as ArrayProxy<any>;
```
Comment (NullVoxPopuli) [inline addon/classes/ChangesetArrayProxyHandler.ts:14]: these should either by symbol properties, or private properties, as they may conflict with folks data

File: addon/classes/ChangesetArrayProxyHandler.ts [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+      this.__source = (source as ArrayProxy<any>).toArray();
+    } else {
+      // wrap the array so it's tracked
+      this.__source = A(source as []);
+    }
+  }
+
+  public get(_target: T, key: string): any {
+    // extra stuff
+    switch (key) {
```
Comment (NullVoxPopuli) [inline addon/classes/ChangesetArrayProxyHandler.ts:24]: you can shorten this by doing `if (typeof this.readArray[key] === 'function') return this.readArray[key];`, etc

File: addon/classes/ChangesetArrayProxyHandler.ts [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+      return this.__nestedProxies.get(index);
+    } else {
+      return this.readArray[index];
+    }
+  }
+
+  @action
+  public getValue(key: string | symbol) {
+    switch (key) {
+      // backwards compatible
```
Comment (NullVoxPopuli) [inline addon/classes/ChangesetArrayProxyHandler.ts:126]: yay back compat

File: addon/utils/is-unchanged.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+export default function isUnchanged(a: any, b: any) {
```
Comment (NullVoxPopuli) [inline addon/utils/is-unchanged.ts:1]: I'm loving all these utilities

**Comment by NullVoxPopuli** [comment]
So, I like where this is going, but the real implementation is here: https://github.com/validated-changeset/validated-changeset
is there a reason you opted to only update this library? or is there a PR I'm missing? :thinking:

File: addon/classes/ChangesetArrayProxyHandler.ts (hunk 1 above)
Comment (BryanCrotaz) [reply addon/classes/ChangesetArrayProxyHandler.ts:14]: you mean __sourceProxy etc?

File: addon/classes/ChangesetArrayProxyHandler.ts (hunk 2 above)
Comment (BryanCrotaz) [reply addon/classes/ChangesetArrayProxyHandler.ts:24]: If you notice I have two different ways of doing it
readArray[key] returns the proxy or the source if there's no changes
writeArray[key] creates a proxy if it doesn't exist and returns it

I thought about using a map to choose which one to return. Otherwise `readKeys.includes[key]` etc

Thoughts?

File: addon/classes/ChangesetArrayProxyHandler.ts (hunk 1 above)
Comment (BryanCrotaz) [reply addon/classes/ChangesetArrayProxyHandler.ts:14]: I don't think these do conflict. Proxy is not a wrapper that falls back to the internals like Ember.ObjectProxy. It explicitly calls in with `get` and `set` and you do your own fallback. __sourceProxy is private api so it's not available through get and set, and if there was a name conflict it would fall through to the proxied object correctly.

**Comment by NullVoxPopuli** [comment]
> validated-changeset is supposed to be independent of Ember

is true -- but it's also where _most_ of the complexity comes in.

is the goal to eliminate the need for validated-changeset?

and fwiw, I think it _is_ possible to a better native proxy design with validated-changeset _and_ provide deep reactivity as well

**Comment by BryanCrotaz** [comment]
My goals are to have ember-changes et working for deep nesting without get and set. If that means submitting to validated- changeset instead then that's fine by me

**Comment by NullVoxPopuli** [comment]
Ah ok, we gan probably do even less work then. Thanks for clarifying!

File: addon/classes/ChangesetArrayProxyHandler.ts (hunk 2 above)
Comment (BryanCrotaz) [reply addon/classes/ChangesetArrayProxyHandler.ts:24]: `typeof this.readArray[key] === 'function'` doesn't work - typescript expects the index to be a number. Is there a way to iterate methods on an array without [key]?

Comment (NullVoxPopuli) [reply addon/classes/ChangesetArrayProxyHandler.ts:24]: I aw not sure

File: addon/classes/ChangesetArrayProxyHandler.ts [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+    // or if a nested proxy is dirty
+    return (this.__proxyArray !== undefined);
+  }
+
+  public get isPristine(): boolean {
+    return (this.__proxyArray === undefined);
+  }
+
+  public get isValid(): boolean {
+    // TODO: validate this level
```
Comment (NullVoxPopuli) [inline addon/classes/ChangesetArrayProxyHandler.ts:94]: Todo!

File: addon/classes/ChangesetArrayProxyHandler.ts [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+
+    // now look at all the nested proxies
+    for (var proxy of this.__nestedProxies.values()) {
```
Comment (NullVoxPopuli) [inline addon/classes/ChangesetArrayProxyHandler.ts:97]: Why reimplement these behaviors? Couldn't you return tte value from the underlying validated-changeset?

File: addon/classes/ChangesetObjectProxyHandler.ts [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+      case 'data':
+      case 'isChangeset':
+      case 'isDirty':
+      case 'isPristine':
+      case 'isValid':
+      case 'pendingData':
+        throw `changeset.${key} is readonly`;
+    }
+    // nested keys are separated by dots
+    var [localKey, subkey] = splitKey(key as string);
```
Comment (NullVoxPopuli) [inline addon/classes/ChangesetObjectProxyHandler.ts:184]: What about deeply nested keys? Maybe 5+ layers deep?

File: addon/classes/ChangesetObjectProxyHandler.ts [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+      }
+      return true;
+    }
+  }
+
+  private markChange(localKey: string, value: any) {
+    // have to use get() here because source might be an EmberProxy
+    const oldValue = get(this.__source, localKey);
+    const unchanged = isUnchanged(value, oldValue);
+    let changes = this.__changes;
```
Comment (NullVoxPopuli) [inline addon/classes/ChangesetObjectProxyHandler.ts:222]: Wait, does this PR remove and re-implement validated-changeset?

Sorry if obvious

**Comment by BryanCrotaz** [comment]
closing in favour of https://github.com/validated-changeset/validated-changeset/pull/150
